### PR TITLE
more sophisticated cookie parsing. Correctly handles multiple cookies

### PR DESCRIPTION
**Pull request by jreinert**
The cookie parsing didn't handle Cookie headers with multiple cookies correctly. Set-Cookie and Cookie headers need to be parsed differently for this. If someone has an idea how to move the regex building into a macro please say so. I couldn't get regex interpolation to work in macros.

### CODE REVIEW

File: src/http/cookie.cr [hunk 1]
```diff
@@ -56,7 +32,75 @@ module HTTP
         header << "; domain=#{domain}" if domain
         header << "; Secure" if @secure
         header << "; HttpOnly" if @http_only
+        header << "; #{@extension}" if @extension
+      end
+    end
+
+    module Parser
```
Comment (jhass) [inline src/http/cookie.cr]: This probably should be a `# :nodoc:` or the methods exposed here should at least point out that the primary interface is `Cookies.from_headers`

File: src/http/cookie.cr [hunk 2]
```diff
@@ -56,7 +32,75 @@ module HTTP
...
+      end
+
+      CookieString = /(?:^|; )#{Regex::CookiePair}/
+      SetCookieString = /^#{Regex::CookiePair}(?:; #{Regex::CookieAV})*$/
+
+      def parse_cookies(header)
+        header.scan(/\s*#{CookieString}\s*/).each do |pair|
+          name = pair["name"]
+          value = pair["value"]
+          yield Cookie.new(name, value)
```
Comment (jhass) [inline src/http/cookie.cr]: Not sure we need the locals here.

File: src/http/cookie.cr [hunk 3]
```diff
@@ -56,7 +32,75 @@ module HTTP
...
+        MaxAgeAV = /Max-Age=(?<max_age>[1-9]*)/i
+        ExpiresAV = /Expires=(?<expires>#{SaneCookieDate})/i
+        CookieAV = /(?:#{ExpiresAV}|#{MaxAgeAV}|#{DomainAV}|#{PathAV}|#{SecureAV}|#{HttpOnlyAV}|#{ExtensionAV})/
+      end
+
+      CookieString = /(?:^|; )#{Regex::CookiePair}/
+      SetCookieString = /^#{Regex::CookiePair}(?:; #{Regex::CookieAV})*$/
+
+      def parse_cookies(header)
+        header.scan(/\s*#{CookieString}\s*/).each do |pair|
```
Comment (jhass) [inline src/http/cookie.cr]: Since `CookieString` doesn't consume whitespace, the `\s*` are redundant, no?

File: src/http/cookie.cr [hunk 4]
```diff
@@ -56,7 +32,75 @@ module HTTP
...
         header << "; Secure" if @secure
         header << "; HttpOnly" if @http_only
+        header << "; #{@extension}" if @extension
+      end
+    end
+
+    module Parser
+      module Regex
+        Token = /[^()<>@,;:\\"\/\[\]?={} \t\x00-\x1f\x7f]+/
+        CookieName = /#{Token}/
```
Comment (jhass) [inline src/http/cookie.cr]: Just `CookieName = Token`?

File: src/http/cookie.cr [hunk 5]
```diff
@@ -56,7 +32,75 @@ module HTTP
...
+        CookieValue = /(?:"#{CookieOctet}*"|#{CookieOctet}*)/
+        CookiePair = /(?<name>#{CookieName})=(?<value>#{CookieValue})/
+        DomainLabel = /[A-Za-z0-9\-]+/
+        Time = /(?:\d{2}:\d{2}:\d{2})/
+        Month = /(?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)/
+        Weekday = /(?:Monday|Tuesday|Wednesday|Thursday|Friday|Saturday|Sunday)/
+        Wkday = /(?:Mon|Tue|Wed|Thu|Fri|Sat|Sun)/
+        PathValue = /[^\x00-\x1f\x7f;]+/
+        DomainValue = /(?:#{DomainLabel}(?:\.#{DomainLabel})?)+/
+        RFC1036Date = /#{Weekday}, \d{2}-#{Month}-\d{2} #{Time} GMT/
```
Comment (jhass) [inline src/http/cookie.cr]: Sure this is forced to be GMT?

**Comment by jhass** [comment]
https://travis-ci.org/manastech/crystal/jobs/91109016#L113

**Comment by jreinert** [comment]
yeah the spec is incorrect. A Cookie header can't have a path attribute

File: src/http/cookie.cr (hunk 5 above)
Comment (jreinert) [inline src/http/cookie.cr]: yup, i built the regex in strict compliance of the RFC.

Comment (jhass) [inline src/http/cookie.cr]: https://tools.ietf.org/html/rfc1036#section-2.1.2 Not this one then.

Comment (jreinert) [inline src/http/cookie.cr]: https://tools.ietf.org/html/rfc2616#section-3.3.1

> All HTTP date/time stamps MUST be represented in Greenwich Mean Time
> (GMT), without exception. For the purposes of HTTP, GMT is exactly
> equal to UTC (Coordinated Universal Time). This is indicated in the
> first two formats by the inclusion of "GMT" as the three-letter
> abbreviation for time zone, and MUST be assumed when reading the
> asctime format. HTTP-date is case sensitive and MUST NOT include
> additional LWS beyond that specifically included as SP in the
> grammar.

Comment (jhass) [inline src/http/cookie.cr]: Ok.

File: spec/std/http/cookie_spec.cr [hunk 6]
```diff
@@ -1,99 +1,129 @@
...
+  end
+end
+
+macro parse_set_cookie(header)
+  begin
+    cookie = HTTP::Cookie::Parser.parse_set_cookie({{header}})
+    cookie.should_not be_nil
+    cookie.not_nil!
+  end
+end
```
Comment (jhass) [inline spec/std/http/cookie_spec.cr]: Actually, why not just make these two methods?

Comment (jreinert) [inline spec/std/http/cookie_spec.cr]: yeah why not

**Comment by jhass** [comment]
Thanks!

**Pull request merged**